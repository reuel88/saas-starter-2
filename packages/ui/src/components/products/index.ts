export { Product } from "./product";
export { Products } from "./products";
export { ProductLoader } from "./product-loader";
