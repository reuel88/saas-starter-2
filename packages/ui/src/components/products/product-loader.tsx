import { FC } from "react";
import { Card, CardFooter, CardHeader } from "../card";
import { Skeleton } from "../ui/skeleton";

export type ProductLoaderProps = {};

export const ProductLoader: FC<ProductLoaderProps> = () => {
  return (
    <Card>
      {/* Image */}
      <Skeleton className="aspect-square" />

      <CardHeader className="space-y-0">
        {/* Title */}
        <div className="py-1">
          <Skeleton className="h-4" />
        </div>

        {/* Price & Rating */}
        <div className="py-[5px]">
          <Skeleton className=" h-[18px]" />
        </div>
      </CardHeader>

      {/* CTA */}
      <CardFooter>
        <Skeleton className="h-10 w-full" />
      </CardFooter>
    </Card>
  );
};
