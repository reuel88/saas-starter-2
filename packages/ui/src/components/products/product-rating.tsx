import { FC } from "react";
import { MdStar } from "react-icons/md";
import { cn } from "@repo/utilities";

type ProductRatingProps = Readonly<{
  rating: number;
}>;

export const ProductRating: FC<ProductRatingProps> = ({ rating }) => {
  const positiveRating = rating >= 0 ? rating : 0;
  const ratingRounded = Math.round(positiveRating);

  return (
    <div className="flex">
      {new Array(5).fill("").map((_, index) => (
        <MdStar
          key={index}
          className={cn(
            "h-5 w-5",
            index < ratingRounded ? "text-hue-color" : "text-muted",
          )}
        />
      ))}
      <span className="text-muted-foreground">({rating})</span>
    </div>
  );
};
