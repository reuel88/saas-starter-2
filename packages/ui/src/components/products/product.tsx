import { FC } from "react";
import { Card, CardFooter, CardHeader, CardTitle } from "../card";
import { Button } from "../button";
import { Carousel, CarouselContent, CarouselItem } from "../ui/carousel";
import { ProductRating } from "./product-rating";

export type ProductProps = {
  title: string;
  price: number;
  rating: number;
  images?: { imageSrc: string }[];
};

export const Product: FC<ProductProps> = ({ title, price, rating, images }) => {
  return (
    <Card as="article">
      <Carousel>
        <CarouselContent>
          {images?.map((image, index) => (
            <CarouselItem key={index}>
              <img
                src={image.imageSrc}
                alt={title}
                className="aspect-square rounded-t-lg object-cover"
              />
            </CarouselItem>
          ))}
        </CarouselContent>
      </Carousel>

      <CardHeader className="space-y-0">
        <CardTitle className="line-clamp-2 text-base font-normal text-muted-foreground">
          {title}
        </CardTitle>
        <div className="flex items-center justify-between">
          <p className="text-lg font-bold">
            {new Intl.NumberFormat("en-AU", {
              style: "currency",
              currency: "AUD",
              minimumFractionDigits: 0,
            }).format(price)}
          </p>
          <div>
            <ProductRating rating={rating} />
          </div>
        </div>
      </CardHeader>
      <CardFooter as="footer">
        <Button type="button" className="w-full">
          Add to Cart
        </Button>
      </CardFooter>
    </Card>
  );
};
