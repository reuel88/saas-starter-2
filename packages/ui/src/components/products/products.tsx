import { FC } from "react";
import { Product, ProductProps } from "./product";

type ProductsProps = {
  products: ProductProps[];
};

export const Products: FC<ProductsProps> = ({ products }) => {
  return (
    <section>
      <header className="flex flex-col">
        <h2 className="text-xl">Products</h2>
        <h3 className="text-muted-foreground">
          <strong className="text-hue-color">10 products</strong> in{" "}
          <strong className="text-hue-color">Categories</strong>
        </h3>
      </header>
      <div className="grid grid-cols-2 gap-4 py-8 lg:grid-cols-3">
        {products.map((product, index) => (
          <Product key={index} {...product} />
        ))}
      </div>
    </section>
  );
};
