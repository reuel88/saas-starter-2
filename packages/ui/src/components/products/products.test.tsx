import { render, screen } from "@testing-library/react";
import { ReactNode } from "react";
import { vi } from "vitest";
import { Products } from "./products";

// Note: The CarouselContent component is mocked to avoid rendering the actual carousel component
// because there is an issue with the carouselRef in CarouselContent.
vi.mock("../ui/carousel", async (importOriginal) => {
  return {
    ...(await importOriginal<typeof import("../ui/carousel")>()),
    CarouselContent: ({ children }: { children: ReactNode }) => {
      return <div>{children}</div>;
    },
  };
});

describe("Products", () => {
  it("should display the title, price, rating, and images of each product", () => {
    // Arrange
    const products = [
      {
        title: "Product 1",
        price: 10,
        rating: 4,
        images: [{ imageSrc: "image1.jpg" }],
      },
      {
        title: "Product 2",
        price: 20,
        rating: 3,
        images: [{ imageSrc: "image2.jpg" }],
      },
    ];

    // Act
    const { baseElement } = render(<Products products={products} />);

    // Assert
    expect(screen.getByText("Product 1")).toBeInTheDocument();
    expect(screen.getByText("$10")).toBeInTheDocument();
    expect(screen.getByAltText("Product 1")).toBeInTheDocument();
    expect(screen.getByText("Product 2")).toBeInTheDocument();
    expect(screen.getByText("$20")).toBeInTheDocument();
    expect(screen.getByAltText("Product 2")).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });
});
