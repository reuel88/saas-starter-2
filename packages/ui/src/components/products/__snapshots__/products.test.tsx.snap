// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Products > should display the title, price, rating, and images of each product 1`] = `
<body>
  <div>
    <section>
      <header
        class="flex flex-col"
      >
        <h2
          class="text-xl"
        >
          Products
        </h2>
        <h3
          class="text-muted-foreground"
        >
          <strong
            class="text-hue-color"
          >
            10 products
          </strong>
           in
           
          <strong
            class="text-hue-color"
          >
            Categories
          </strong>
        </h3>
      </header>
      <div
        class="grid grid-cols-2 gap-4 py-8 lg:grid-cols-3"
      >
        <article
          class="rounded-lg border bg-card text-card-foreground shadow-sm"
        >
          <div
            aria-roledescription="carousel"
            class="relative"
            role="region"
          >
            <div>
              <div
                aria-roledescription="slide"
                class="min-w-0 shrink-0 grow-0 basis-full pl-4"
                role="group"
              >
                <img
                  alt="Product 1"
                  class="aspect-square rounded-t-lg object-cover"
                  src="image1.jpg"
                />
              </div>
            </div>
          </div>
          <div
            class="flex flex-col p-6 space-y-0"
          >
            <h3
              class="tracking-tight line-clamp-2 text-base font-normal text-muted-foreground"
            >
              Product 1
            </h3>
            <div
              class="flex items-center justify-between"
            >
              <p
                class="text-lg font-bold"
              >
                $10
              </p>
              <div>
                <div
                  class="flex"
                >
                  <svg
                    class="h-5 w-5 text-hue-color"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    class="h-5 w-5 text-hue-color"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    class="h-5 w-5 text-hue-color"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    class="h-5 w-5 text-hue-color"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    class="h-5 w-5 text-muted"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <span
                    class="text-muted-foreground"
                  >
                    (
                    4
                    )
                  </span>
                </div>
              </div>
            </div>
          </div>
          <footer
            class="flex items-center p-6 pt-0"
          >
            <button
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full"
              type="button"
            >
              Add to Cart
            </button>
          </footer>
        </article>
        <article
          class="rounded-lg border bg-card text-card-foreground shadow-sm"
        >
          <div
            aria-roledescription="carousel"
            class="relative"
            role="region"
          >
            <div>
              <div
                aria-roledescription="slide"
                class="min-w-0 shrink-0 grow-0 basis-full pl-4"
                role="group"
              >
                <img
                  alt="Product 2"
                  class="aspect-square rounded-t-lg object-cover"
                  src="image2.jpg"
                />
              </div>
            </div>
          </div>
          <div
            class="flex flex-col p-6 space-y-0"
          >
            <h3
              class="tracking-tight line-clamp-2 text-base font-normal text-muted-foreground"
            >
              Product 2
            </h3>
            <div
              class="flex items-center justify-between"
            >
              <p
                class="text-lg font-bold"
              >
                $20
              </p>
              <div>
                <div
                  class="flex"
                >
                  <svg
                    class="h-5 w-5 text-hue-color"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    class="h-5 w-5 text-hue-color"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    class="h-5 w-5 text-hue-color"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    class="h-5 w-5 text-muted"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    class="h-5 w-5 text-muted"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <span
                    class="text-muted-foreground"
                  >
                    (
                    3
                    )
                  </span>
                </div>
              </div>
            </div>
          </div>
          <footer
            class="flex items-center p-6 pt-0"
          >
            <button
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full"
              type="button"
            >
              Add to Cart
            </button>
          </footer>
        </article>
      </div>
    </section>
  </div>
</body>
`;
